Add switch to connected account alert

Gudahtt opened the pull request:
Add alert suggesting that the user switch to a connected account. This alert is displayed when the popup is opened over an active tab that is connected to some account, but not the current selected account. The user can choose to switch to a connected account, or dismiss the alert.

This alert is only shown once per account switch. So if the user repeatedly opens the popup on a dapp without switching accounts, it'll only be shown the first time.

A checkbox allowing the user to dismiss the alert permanently will be added in a subsequent PR.

Gudahtt committed:
Add switch to connected account alert

Add alert suggesting that the user switch to a connected account. This
alert is displayed when the popup is opened over an active tab that is
connected to some account, but not the current selected account. The
user can choose to switch to a connected account, or dismiss the alert.

This alert is only shown once per account switch. So if the user
repeatedly opens the popup on a dapp without switching accounts, it'll
only be shown the first time.

A checkbox allowing the user to dismiss the alert permanently will be
added in a subsequent PR.

Gudahtt commented:
This depends upon #8524

Gudahtt commented:
I need to check on a couple of more things before this is ready to review.

* The check for whether the alert was shown already isn't per-dapp right now, but it should be.
* There's a scrollbar that keeps showing up on the home screen that I want to get rid of.

Gudahtt committed:
Track shown state per-domain

Gudahtt commented:
<details>
<summary>Screenshots:</summary>
Disregard the scrollbar in these two screenshots - that was removed in a separate PR (#8547)

Multiple accounts:

![switch-to-alert](https://user-images.githubusercontent.com/2459287/81301058-ed38e880-904e-11ea-8c0e-2341d5c0bc4f.png)

One account:

![switch-to-alert-1-account](https://user-images.githubusercontent.com/2459287/81301072-f32ec980-904e-11ea-9768-cb544a54a127.png)


</details>

Gudahtt commented on app/scripts/controllers/app-state.js:
```diff
@@ -107,6 +121,16 @@ export default class AppStateController extends EventEmitter {
+   * Sets the "switch to connected" alert as shown for the given origin
+   * @param {string} origin - The origin the alert has been shown for
+   */
+  setSwitchToConnectedAlertShown (origin) {
```
After discussing this with @rachelcope , we've decided to set this alert to "shown" after the user sees the _other_ alert too (the "Unconnected account" alert seen after switching accounts in the popup). This is to prevent them from seeing two similar alerts in a row.

An example of this case is:

* The user opens the popup on a dapp, with a connected account selected
* The user switches to an unconnected account, sees the "unconnected account" alert, then dismisses it
* The user closes the popup
* The user opens the popup again

In that case, currently we'd show the "switch to account" alert. This is what we'd like to change, so that this alert is skipped in this specific scenario.

Gudahtt replied:
This has been done in 6b320b70505425a342bb9268f1f180e2da79464e

Gudahtt committed:
Set alert as shown when showing similar alert

The "Switch to connceted" alert is now set as "shown" after the
"unconnected account" alert has been shown. This is to prevent the user
from being shown two very similar alerts in a short span of time.

Gudahtt commented on app/scripts/controllers/app-state.js:
```diff
@@ -15,28 +15,42 @@ export default class AppStateController extends EventEmitter {
+        },
+        initState,
+        {
+          switchToConnectedAlertShown: {},
```
I think this PR is fine to merge as-is, but it might be better to move this stuff into the new `AlertController` created in #8550. If so, maybe it'd be better to get #8550 merged first. Either way I move this afterwards in a separate PR though.

Gudahtt replied:
This was done in 55f4b3e22e2d2fd63d92dcce09b5b570d875e9fd

rachelcope reviewed (dismissed):
One small design feedback - looks like there is too much left/right padding on the input (making it inconsistently aligned with the title and buttons). 
Otherwise, LGTM!

rachelcope commented:
Should this have the "don't show this again" functionality?

Gudahtt commented:
>Should this have the "don't show this again" functionality?

Nope! Though I'm considering updating this PR to include it before merging, now that #8550 is pretty close to being merged.

Gudahtt committed:
Update padding around account selection input

Gudahtt commented:
> looks like there is too much left/right padding on the input (making it inconsistently aligned with the title and buttons).

Good catch! I've just updating the input padding to match the text.

<details>
<summary>Here's what it looks like now:</summary>

![updated-padding](https://user-images.githubusercontent.com/2459287/81436836-7f6ce980-9140-11ea-929b-294a9f4a0e91.png)

</details>

Gudahtt commented:
I've also just realized that I handled the single-account case wrong. It's supposed to look like this if there's one account:

<details>
<summary>Screenshot: </summary>

![switch-accounts](https://user-images.githubusercontent.com/2459287/81436959-bba04a00-9140-11ea-8f20-56d0527908e5.PNG)

</details>

Gudahtt committed:
Switch to using new ALERT_TYPES switchToConnected enum

Gudahtt committed:
Move 'switchToConnectedAlertShown' from app state controller to alert controller

Gudahtt committed:
Add 'don't show this again' checkbox to alert

Gudahtt commented:
The "don't show this again" checkbox has been added to this alert, along with the associated alert settings.

<details>
<summary>Screenshots:</summary>

Alert with checkbox:

![with-checkbox](https://user-images.githubusercontent.com/2459287/81446977-25751f80-9152-11ea-947b-6df1e400a20f.png)

Updated alert settings:

![alert-settings](https://user-images.githubusercontent.com/2459287/81446989-2d34c400-9152-11ea-98f0-f327c0b937d5.png)

</details>

Gudahtt committed:
Update copy, update single-account design

The alert has been updated to omit the dropdown when there is just a
single connected account. The copy for both the single and multiple
account states has also been updated.

Gudahtt commented:
>I've also just realized that I handled the single-account case wrong.

After discussing this with @rachelcope , the copy for both the single and multiple-account cases has been updated. The alert has also been updated to omit the dropdown if there is just a single account.

<details>
<summary>Here are the updated screenshots:</summary>

Single account:

![new-one-account](https://user-images.githubusercontent.com/2459287/81458366-64b26900-9170-11ea-8041-b02c231b2468.png)

Multiple accounts:

![new-multiple-accounts](https://user-images.githubusercontent.com/2459287/81458370-6c720d80-9170-11ea-9a85-f63ca958c562.png)

</details>

Gudahtt commented on ui/app/ducks/alerts/unconnected-account.js:
```diff
@@ -94,7 +94,7 @@ export const dismissAndDisableAlert = () => {
     try {
       await dispatch(disableAlertRequested())
-      await dispatch(setAlertEnabledness(name), false)
+      await dispatch(setAlertEnabledness(name, false))
```
I didn't intend to commit this here - this has been fixed in a separate PR (#8561)

Gudahtt replied:
#8561 has been merged, and this PR has been rebased on `develop`, so this should be gone from the diff now.

whymarrh reviewed (approved):
LGTM!
